src/main/kotlin/cacophony/codegen/registers/SpillHandling.kt

```diff
@@ -0,0 +1,105 @@
+ * @return List of BasicBlocks with adjusted and new instructions.
+ */
+fun adjustLoweredCFGToHandleSpills(
+    instructionCovering: InstructionCovering,
+    functionHandler: FunctionHandler,
+    loweredCfg: LoweredCFGFragment,
+    registerAllocation: RegisterAllocation,
+    spareRegisters: Set<FixedRegister>,
+): LoweredCFGFragment {
+    if (registerAllocation.successful.values
+            .intersect(spareRegisters.map { it.hardwareRegister }.toSet())
+            .isNotEmpty()
+    ) {
+        throw SpillHandlingErrorException("Spare register is used in register allocation.")
+    }
+
+    if (registerAllocation.spills.isEmpty()) {
+        return loweredCfg
+    }
+
+    val spills: Set<VirtualRegister> =
+        registerAllocation.spills
+            .map {
+                when (it) {
+                    is FixedRegister -> throw SpillHandlingErrorException("Fixed register $it has spilled.")
+                    is VirtualRegister -> it
+                }
+            }.toSet()
+
+    val spillsFrameAllocation: Map<VirtualRegister, CFGNode.LValue> =
+        spills.associateWith {
+            functionHandler.allocateFrameVariable(it)
+        }
+
+    fun loadSpillIntoReg(spill: VirtualRegister, reg: FixedRegister): List<Instruction> =
+        instructionCovering.coverWithInstructions(
+            CFGNode.Assignment(
+                CFGNode.RegisterUse(reg),
+                spillsFrameAllocation[spill]!!,
+            ),
+        )
+
+    fun saveRegIntoSpill(reg: FixedRegister, spill: VirtualRegister): List<Instruction> =
+        instructionCovering.coverWithInstructions(
+            CFGNode.Assignment(
+                spillsFrameAllocation[spill]!!,
+                CFGNode.RegisterUse(reg),
+            ),
+        )
+
+    return loweredCfg.map { block ->
+        val newInstructions =
+            block.instructions().flatMap { instruction ->
+                val readSpills = instruction.registersRead.filterIsInstance<VirtualRegister>().filter { spills.contains(it) }
+                val writtenSpills = instruction.registersWritten.filterIsInstance<VirtualRegister>().filter { spills.contains(it) }
+                val usedSpills = readSpills + writtenSpills
+
+                if (usedSpills.size > spareRegisters.size) {
+                    throw SpillHandlingErrorException(
+                        "Not enough spare registers: Detected instruction with ${usedSpills.size} registers," +
+                            "but only have ${spareRegisters.size} spare registers to use",
+                    )
+                }
+
+                val registersSubstitution: Map<Register, FixedRegister> = usedSpills.zip(spareRegisters).toMap()
+
+                val instructionPrologue = readSpills.flatMap { loadSpillIntoReg(it, registersSubstitution[it]!!) }
+                val instructionEpilogue = writtenSpills.flatMap { saveRegIntoSpill(registersSubstitution[it]!!, it) }
+
+                instructionPrologue +
```

Comment:
Can you fix this formatting or is it forced by a linter?

Reply:
fixed